packages/nimble-components/src/rich-text-mention/base/models/mention-validator.ts

```diff
@@ -0,0 +1,109 @@
+    'missingPatternAttribute'
+] as const;
+
+/**
+ * Validator for RichTextMention
+ */
+export class RichTextMentionValidator<
+    ValidityFlagNames extends readonly string[]
+> extends Validator<typeof baseValidityFlagNames | ValidityFlagNames> {
+    public constructor(
+        private readonly mentionInternals: MentionInternals<unknown>,
+        configValidityKeys: ValidityFlagNames,
+        private readonly supportedMappingElements: readonly (typeof Mapping<unknown>)[]
+    ) {
+        super(configValidityKeys);
+    }
+
+    public validate(
+        mappings: Mapping<unknown>[],
+        pattern: string | undefined
+    ): void {
+        this.untrackAll();
+        const mentionHrefs = mappings.map(mapping => mapping.key);
+        this.validateMappingTypes(mappings);
+        this.validateNoMissingMentionHref(mappings);
+        this.validateUniqueMentionHref(mentionHrefs);
+        this.validatePattern(pattern);
+        this.validateHref(mentionHrefs, pattern);
+    }
+
+    /**
+     * @returns an object containing flags for various ways the configuation can be invalid
+     */
+    public getValidity(): RichTextMentionValidity {
+        return this.getValidationFlags();
+    }
+
+    /**
+     * Sets a particular validity condition flag's value, e.g. "hasInvalidFooValue" = true
+     */
+    protected setConditionValue(
+        name: FlagNames<typeof baseValidityFlagNames | ValidityFlagNames>,
+        isInvalid: boolean
+    ): void {
+        if (isInvalid) {
+            this.track(name);
+        } else {
+            this.untrack(name);
+        }
+        this.updateMentionInternalsFlag();
+    }
+
+    private updateMentionInternalsFlag(): void {
+        this.mentionInternals.validConfiguration = this.isValid();
+    }
+
+    private validatePattern(pattern: string | undefined): void {
+        this.setConditionValue(
+            'missingPatternAttribute',
+            pattern === undefined
+        );
+    }
+
+    private validateMappingTypes(mappings: Mapping<unknown>[]): void {
+        const valid = mappings.every(mapping => this.supportedMappingElements.some(
+            mappingClass => mapping instanceof mappingClass
+        ));
+        this.setConditionValue('unsupportedMappingType', !valid);
+    }
+
+    private validateUniqueMentionHref(
+        mentionHref: (string | undefined | unknown)[]
+    ): void {
+        const invalid = new Set(mentionHref).size !== mentionHref.length;
+        this.setConditionValue('duplicateMappingMentionHref', invalid);
+    }
+
+    private validateNoMissingMentionHref(mappings: Mapping<unknown>[]): void {
+        const invalid = mappings.some(mapping => mapping.key === undefined);
+        this.setConditionValue('missingMentionHrefValue', invalid);
+    }
+
+    private validateHref(
+        mentionHrefs: unknown[],
+        pattern: string | undefined
+    ): void {
+        const regexPattern = new RegExp(pattern!);
+        const valid = mentionHrefs.every(
```

Comment:
Is this logic correct? It looks like we're saying it's valid to provide an `href` that's `undefined` or not a string. Why would we allow that?

Reply:
The conditions here should be a compliment to those used in `validateNoMissingMentionHref`, i.e. if `validateNoMissingMentionHref` catches `undefined` hrefs, `validateHref` should ignore `undefined` hrefs. However, it does seem like hrefs that are defined but not a string should trigger the `unsupportedMentionHrefValue` flag.

Reply:
As the `test()` method will expect only `string` as input as well as the `url` should be of string type, `href` need to be validated for `string` type. So in that case even a defined `href` of other types will also be validated under `unsupportedMentionHrefValue` flag.